appcues/src/main/java/com/appcues/ui/ExperienceRenderer.kt

```diff
@@ -2,91 +2,109 @@ package com.appcues.ui
     }
 
-    fun process(renderContext: RenderContext, actions: List<ExperienceAction>, interactionType: InteractionType, viewDescription: String?) {
-        actionProcessor.process(renderContext, actions, interactionType, viewDescription)
+    fun getState(renderContext: RenderContext): State {
+        return slots[renderContext]?.state ?: Idling
     }
 
-    suspend fun show(experience: Experience): Boolean {
+    /**
+     * returns true/false whether the experience is was started
+     */
+    suspend fun show(experience: Experience): Boolean = with(experience) {
         var canShow = config.interceptor?.canDisplayExperience(experience.id) ?: true
 
         // if there is an active experiment, and we should not show this experience (control group), then
         // track the analytics for experiment_entered, but ensure we exit early. This should be checked before
         // we dismiss any current experience below.
-        if (experience.experiment != null && !experience.experiment.shouldExecute()) {
-            experience.experiment.track(analyticsTracker)
+        if (experiment?.group == "control") {
+            analyticsTracker.track(experiment)
             canShow = false
         }
 
         if (!canShow) return false
 
+        val stateMachine = slots[renderContext]
+            ?: run {
+                get<StateMachine>()
+                    .also { slots[renderContext] = it }
+                    .also { lifecycleTracker.start(it) }
+            }
+
+        val currentExperience = stateMachine.state.currentExperience
+
+        return if (currentExperience.shouldDismissAndShow(this)) {
+            stateMachine.dismissAndShow(this)
+        } else {
+            // track an experiment_entered analytic, if exists, since we know it is not in the control group at this point
+            experiment?.let { analyticsTracker.track(it) }
+
+            stateMachine.show(this)
+        }
+    }
+
+    private fun Experience?.shouldDismissAndShow(newExperience: Experience): Boolean {
         // "event_trigger" or "forced" experience priority is NORMAL, "screen_view" is low -
         // if an experience is currently showing and the new experience coming in is normal priority
         // then it replaces whatever is currently showing - i.e. an "event_trigger" experience will
         // supersede a "screen_view" triggered experience - per Appcues standard behavior
-        val priorityOverride = experience.priority == NORMAL && stateMachine.state != Idling
-        if (priorityOverride) {
-            return dismiss(RenderContext.Modal, markComplete = false, destroyed = false).run {
-                when (this) {
-                    is Success -> show(experience) // re-invoke show on the new experience now after dismiss
-                    is Failure -> false // dismiss failed - can't continue
-                }
-            }
-        }
+        return this != null && this != newExperience && newExperience.priority == NORMAL
```

Comment:
I thought about it when I was trying to make a test work. not sure if this is ok so I want to confirm.

instead of checking the stateMachine state, we can (I think) check the exiting experience (this) is different than null, and if (this) is different than the incoming experience.

this works well because during tests we init a StateMachine with a specific state like RenderingStep(experience ...) then since its the same the dismiss is ignored keeping that mocked machine on its current state even if the priority of the new experience is set to NORMAL. 

again if this looks wrong we can revert it and make sure that the mocked experience in tests are set to LOW (that works too)

Reply:
I need to wrap my head around this, im not sure. My gut instinct is to retain existing behavior wherever possible to minimize chance of regression, but I don't fully understand the change yet.

Reply:
OK I went through this again and I agree I think it is logically the same. Since the only case where `stateMachine.state.currentExperience` is `null` i the `Idling` case, checking for `this != null` has the same effect as checking `stateMachine.state != Idling` in the previous logic.

I am curious which test exactly needed the change and how it worked before without the change, just to understand that part.

Reply:
thanks for doublechecking on that. I added more context in [here](https://appcues.slack.com/archives/C031RLGS4F8/p1689601592014059)